### ISSUE TITLE
[discussion] signature of move function and Game object for new API clients

### ISSUE THREAD

**Issue by otizonaizit** [ASPP/pelita#529]
Context of the discussion: #520  #522 

The question is that, if we go the route of #527 , the `move` function for a client will look like something like this:

```python
def move(turn, game, state):
    bot = game.team[turn]
    other = game.team[1-turn]
    ...
    return next_move, state
```
Given that with this the `Game` object only contains the attribute `team`, it makes sense to get rid of `game` alltogether and directly pass `team`, as in:

```python
def move(turn, team, state):
    bot = team[turn]
    other = team[1-turn]
    ...
    return next_move, state
```

Open for comments and/or discussion.

**Comment by Debilski** [comment]
Game currently holds out pretty-print code but in principle I agree that it is less useful now.

**Comment by otizonaizit** [comment]
Alternative variant would be:

```python
def move(bot, other, state):
    ....
   return next_move, state
```

The advantage I see here is that we don't neet the `Team` object which is nothing else than a list. On the other hand, if `turn` is not passed explicitly, students will have to come up with their own identification when they want to use `state`. But this could be explained in the docs.

**Comment by otizonaizit** [comment]
Yet another alternative:

```python
def move(bot, state):
    other = bot.other
    ...
    return next_move, state
```
This is somehow the one I like more, but I am not sure if it can create confusion. Looking at `pelita_template` it seems to me that this would simplify the client code a lot. And we get rid of two objects: docs are more straightforward and there is no discussion anymore where do things belong to: everything is in bot...

**Comment by Debilski** [comment]
But `setup_test_game` would return the list of your bots? Or would you specify which one with a separate argument? I don’t see the list as that much of a problem because it corresponds to the layout specification in the tests.

I’ll have to think about it a little. I think there must have been a reason why we didn’t do it like this in the first place.

**Comment by otizonaizit** [comment]
> But `setup_test_game` would return the list of your bots? Or would you specify which one with a separate argument? I don’t see the list as that much of a problem because it corresponds to the layout specification in the tests.

The `setup_test_game` will return just `bot`. That bot will be always the `0`. That seems the most straightforward thing to do. The list is a problem in as much as it is not needed anymore. The layout specifications in the test are surely not something we want to use to orient ourselves for designing an API that feels intuitive. How to write the tests can be demonstrated in the demos. But again, to me it seems your objections are of esthetic character and somehow realted to the fact that you know how it was implemented before (see your quote below).
 
> I’ll have to think about it a little. I think there must have been a reason why we didn’t do it like this in the first place.

The reason is because we were still thinking coming from the old API and without any experience how the new API works in the wild. Now we know, and to me it seems clear that flattening the hierarchy more makes the last confusions go away.

**Comment by Debilski** [comment]
Passing the bot as the first argument *is* the old API though …

**Comment by otizonaizit** [comment]
No, it's not. the old API required you to subclass `AbstracPlayer` etc.

**Comment by otizonaizit** [comment]
I had a go at seeing how does it feel to write code using my proposed third alternative:
```python
def move(bot, state):
    next_move = (0,0)
    return next_move, state
```
See the results at https://github.com/ASPP/pelita_template/pull/9

I think it's worth it.

**Comment by otizonaizit** [comment]
In particular, `demo06_one_and_one.py` is so much nicer and straightforward.

**Comment by otizonaizit** [comment]
By the way, I called the bot index `bot.turn`. In the documentation it reads way more natural than `bot.index`.

**Comment by Debilski** [comment]
`Game` is gone from https://github.com/ASPP/pelita/pull/527 and we have too many discussion threads for basically the same thing.

**Comment by Debilski** [comment]
I am not the biggest fan of `bot.other` but I don’t really have a better suggestion for now.

Other open questions: Should we revive `bot.team` again (it is currently `_team`) because now it is not quite as redundant anymore as before (Before we had `game.team` for the same thing)?

I suggest we make `Bot` unhashable to avoid having people do `state[bot] = "attacker"` (unless we find a smart hash that would allow them to do it).

**Comment by jni** [comment]
> I am not the biggest fan of bot.other but I don’t really have a better suggestion for now.

`bot.teammate`?

> Should we revive bot.team again?

Is this roughly equivalent to `[bot, bot.teammate]`?

**Comment by otizonaizit** [comment]
> > I am not the biggest fan of bot.other but I don’t really have a better suggestion for now.
> 
> `bot.teammate`?

Yes, `bot.teammate` or `bot.other`, I prefer the shortest one, and at the end is only a matter of documentation, it's not like there is much space for confusion in the new API. Look at the use of `bot.other` in  https://github.com/ASPP/pelita_template/tree/state-as-an-arg : it feels natural and understandable to me.

**Comment by otizonaizit** [comment]
> Other open questions: Should we revive `bot.team` again (it is currently `_team`) because now it is not quite as redundant anymore as before (Before we had `game.team` for the same thing)?

Why are you re-asking this question again? Why now? I don't get it. We are experimenting with having only `bot` and `bot.other`, we have a proof of concept of how to use the new API in   https://github.com/ASPP/pelita_template/tree/state-as-an-arg . Just look at the code there: is there any place there where you think having `bot.team` would be useful and/or more clear than what we have now?
If not, I do not get why we are to discuss about `bot.team` again.
And no, a "special" `Team` object, which is anything more than a list of bots is even less meaningful at this point in the discussion...

**Comment by otizonaizit** [comment]
> make `Bot` unhashable to avoid having people do `state[bot] = "attacker"` (unless we find a smart hash that would allow them to do it).

Yes, that is a good point. Make it unhashable. They can do `state[bot.turn]` and there is already an example in the demos doing it, so the functionality is there.

**Comment by Debilski** [comment]
> Why are you re-asking this question again? Why now?

Well, of course now. That’s the point of this thread, no? ;)

`bot.team` would not be a special team object but simply the list of the two bots, mirroring `bot.enemy`. Of course, we can argue that `[bot, bot.other]` serves the same purpose but I’m not sure. Having `bot.team` available is arguably less error prone in some situations than always having to build it yourself.

Some of the groups also have been using the old `game.team` object together with indices for storing state. To me that is a stronger argument than our own template code.
The fact that it is ordered by id makes certain things easier, when you just want to update state. If you write your own tracker, you don’t care about ‘bot’ or ‘bot.other’ you want to have all of them and in a stable order. In our template code, we don’t really run into the same situation, because our example bots are mostly written without the whole team in mind, so I’d say it’s not really comparable.

An additional minor reason to keep it is that we are using `bot._team` in the internal API for the pretty print and I don’t see a reason why we shouldn’t just document and make it public then. If people are doing `dir(bot)` they’ll find out about it anyway.

**Comment by Debilski** [comment]
For reference. My cryptic comment here (https://github.com/ASPP/pelita/issues/529#issuecomment-423731195) meant: `Game` has been removed from the code and everything is implemented in #527 as requested with the new move API.

**Comment by otizonaizit** [comment]
This is done.
